### PR TITLE
Use parse-error and tc-error helpers consistently

### PR DESCRIPTION
**Pull request by jbouwman**
Use `parse-error` and `tc-error` systematically.

Reverse the argument order to `tc-error` in order to directly reflect how conditions are presented: as a message that applies to the condition as a whole, a primary span (to which an editor ought to jump, when presenting the condition), and optional auxiliary notes and 'help notes'.

The note in src/parse/base.lisp describes usage:

```
;;; A complex parse error may be signaled with:
;;;
;;;   (parse-error "Overall description of condition"
;;;                (note SOURCE CST1 "Primary ~A: ~A" ARG1 ARG2)
;;;                (note SOURCE CST2 "Related: ~A" ARG3)
;;;                ... )
```

- unused condition classes related to warnings are removed from analysis phases

### CODE REVIEW

File: src/typechecker/environment.lisp [hunk 1]
```diff
@@ -513,9 +513,9 @@
   '(satisfies struct-field-list-p))
 
 (defstruct struct-entry
-  (name             (util:required 'name)      :type symbol            :read-only t)
```
Comment (Izaakwltn) [inline src/typechecker/environment.lisp:516]: Maybe move indentation fixes to a separate PR

File: src/typechecker/expression.lisp [hunk 2]
```diff
@@ -205,9 +205,9 @@
 
 (defstruct (node-bind
             (:copier nil))
-  (pattern  (util:required 'pattern)   :type pattern         :read-only t)
-  (expr     (util:required 'expr)      :type node            :read-only t)
```
Comment (Izaakwltn) [inline src/typechecker/expression.lisp:209]: Indentation PR perhaps

File: tests/parser-test-files/lisp-toplevel.txt [hunk 3]
```diff
@@ -2,7 +2,7 @@
 Empty options
 
 To support future extension, lisp-toplevel expects an empty options list in the
-first poisition.
+first position.
```
Comment (Izaakwltn) [inline tests/parser-test-files/lisp-toplevel.txt:5]: Removing flavor from the Coalton cast iron

File: src/typechecker/pattern.lisp [hunk 4]
```diff
@@ -40,8 +40,8 @@
 (defstruct (pattern
             (:constructor nil)
             (:copier nil))
-  (type     (util:required 'type)   :type tc:qualified-ty           :read-only t)
```
Comment (Izaakwltn) [inline src/typechecker/pattern.lisp:43]: Probably indentation PR

**Review by Izaakwltn — APPROVED** [review]
A few thoughts, otherwise LGTM

File: src/typechecker/environment.lisp (hunk 1 above)
Comment (jbouwman) [reply src/typechecker/environment.lisp]: done

File: src/typechecker/expression.lisp (hunk 2 above)
Comment (jbouwman) [reply src/typechecker/expression.lisp]: done

File: src/typechecker/pattern.lisp (hunk 4 above)
Comment (jbouwman) [reply src/typechecker/pattern.lisp]: done
